broadinstitute/scp-ingest-pipeline
# Transform `has_<modality>` to boolean for BigQuery (SCP-5566)

#### BACKGROUND & CHANGES
[Proposed ingest design](https://docs.google.com/presentation/d/1L2ah87lFDvL8gnFP0IRUTBe9wmwmne28iWpJAvTANzA/edit#slide=id.g1ff70cc6193_0_31) for eventual BRAIN RF1/NeMO ingest via Import Service envisions, ingest pipeline anndata "extract" translating metadata column headers like "dandi_url" to "has_electrophysiology". This work applies to the subsequent step of metadata ingest to transform the URL information in "has_<modality>" metadatum to TRUE for BigQuery ingest while keeping the URL information for MongoDB ingest.


#### MANUAL TESTING
1. Modify your BigQuery schema for your local instance manually like so:
![Screenshot 2024-05-06 at 1 20 12 PM](https://github.com/broadinstitute/scp-ingest-pipeline/assets/6005195/e8f46ce3-d68c-434a-be34-13c4b3737042)
a.  Go to https://console.cloud.google.com/bigquery?project=$PROJECT, where $PROJECT has a value like broad-singlecellportal-jlchang
b.  In the Explorer panel at left, click the arrow in your $PROJECT to toggle that node open
c.  Click arrow in "cell_metadata" to open Dataset node
d.  Click "alexandria_convention" (Schema node)
e.  Click blue "EDIT SCHEMA" button at bottom of page
f.  Scroll down to "New fields", click "+" button to add field
g.  Add field name "has_electrophysiology", select type "Boolean"
h.  Add field name "has_morphology", select type "Boolean"
i.  Click "SAVE"
2. Run ingest pipeline via command line
a. Activate the scp-ingest-pipeline repo virtualenv
b. From the ingest directory of the scp-ingest-pipeline repo, perform this setup:
`source ../scripts/setup-mongo-dev.sh <path to your Github token file>`
If you get an authentication error from gcloud, run the following:
`gcloud auth application-default login --no-launch-browser `
c. Run this metadata ingest job from the ingest directory of the scp-ingest-pipeline repo:

> python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 ingest_cell_metadata --cell-metadata-file ../tests/data/annotation/metadata/convention/brain_rf1//patchseq_classic_metadata_has_modality_10.tsv --study-accession SCPPR344 --ingest-cell-metadata --validate-convention --has-modality "['electrophysiology', 'morphology']" --bq-dataset cell_metadata_development --bq-table alexandria_convention

3. In the following BigQuery query, replace \<username\> with your username:

> Select * from `broad-singlecellportal-<username>.cell_metadata.alexandria_convention` where study_accession = "SCPPR344"

4. Confirm that 9 rows are returned with the following boolean metadata:
<img width="348" alt="Screenshot 2024-05-20 at 12 05 32 PM" src="https://github.com/broadinstitute/scp-ingest-pipeline/assets/6005195/c97f0751-e0e6-4d04-98a1-078576c0b751">

5. and the job returns with the following success output:

> Metadata uploaded to BigQuery. (9 rows)
> distinct_id: 2f30ec50-a04d-4d43-8fd1-b136a2045079
> studyAccession: SCPdev
> fileName: dec0dedfeed1111111111111
> fileType: input_validation_bypassed
> fileSize: 1
> trigger: dev-mode
> logger: ingest-pipeline
> appId: single-cell-portal
> status: success
> functionName: ingest_cell_metadata
> perfTime: 0.01

Note: your distinct_id will be different
